omegaconf/grammar/OmegaConfGrammarParser.g4

```diff
@@ -41,7 +41,13 @@ sequence: (element (COMMA element?)*) | (COMMA element?)+;
 // Interpolations.
 
 interpolation: interpolationNode | interpolationResolver;
-interpolationNode: INTER_OPEN DOT* (configKey (DOT configKey)*)? INTER_CLOSE;
+interpolationNode:
+      INTER_OPEN DOT*                                            // dots for relative interpolations
+      (
+        configKey                                                // first config key
+        (DOT configKey | BRACKET_OPEN configKey BRACKET_CLOSE)*  // .foo, [foo], .foo[bar], [foo].bar[baz]
```

Comment:
I think the first can also be bracket index:
```python
lc = OmegaConf.create([
  {"a": 10}, 
  {"b": 20}
])

assert OmegaConf.select(lc, "[0].a") == 10
```

Reply:
So, I thought about this, but I think it may be confusing because `x[0]` is equivalent to `x.0`, but `[0]` wouldn't be equivalent to `.0`
At the same time, I was wondering if we should make `[0]` equivalent to `.0`, and I thought this may not be very intuitive precisely because someone may instead expect the behavior you describe here. So I thought "let's just not support it :)"
But it's a small point so let me know what you prefer.

Reply:
This is in part why get_full_key is such a complicated function (by the way, it uses index notation for lists, but looks like not very consistently (that's a small bug there):
```python
In [24]: cfg = OmegaConf.create("[{'a':10}, 2]")

In [25]: cfg
Out[25]: [{'a': 10}, 2]

In [26]: cfg[0]._get_full_key("")
Out[26]: 0

In [27]: cfg[0]._get_full_key("a")
Out[27]: '[0].a'
```
making [0] equivalent to .0 will make it relative by default: its meaning will also differ from 0, I think that's not good.

I think we should make those equivalent:
```
a == [a]
0 == [0]
foo.bar == [foo].bar == foo[bar]
a.b.c == [a].b[c] == a.b[c]
```
```
# start of key:
[k] == k
# middle of key:
[k] == .k
```

While this may look weird, I think people would not be surprised by the behavior.
The case of a top level list is a bit unusual, but the case with relative interpolations may be more common:
```yaml
list:
  - a:
      b: 10
  - ${0.a}
  - ${[0].a}
```

Reply:
Ok, done in 8131646

Note that your example isn't actually a relative interpolation, here's the correct version (as of the above commit):
```yaml
list:
  - a:
      b: 10
  - ${.0.a}
  - ${.[0].a}
```